api/handlers.go

```diff
@@ -248,31 +279,39 @@ func (api *API) registerGET(w http.ResponseWriter, req *http.Request, _ httprout
 
 // registerPOST registers a new user based on a challenge-response.
 func (api *API) registerPOST(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	// Get the body, we might need to use it several times.
+	body, err := ioutil.ReadAll(io.LimitReader(req.Body, 4*skynet.KiB))
+	if err != nil {
+		api.WriteError(w, errors.AddContext(err, "empty request body"), http.StatusBadRequest)
+		return
+	}
 	// Get the challenge response.
 	var chr database.ChallengeResponse
-	err := chr.LoadFromRequest(req)
+	err = chr.LoadFromBytes(body)
 	if err != nil {
 		api.WriteError(w, errors.AddContext(err, "missing or invalid challenge response"), http.StatusBadRequest)
 		return
 	}
-	ctx := req.Context()
-	pk, _, err := api.staticDB.ValidateChallengeResponse(ctx, chr, database.ChallengeTypeRegister)
+	// Parse the request's body.
+	var payload credentialsPOST
+	err = json.Unmarshal(body, &payload)
 	if err != nil {
-		api.WriteError(w, errors.AddContext(err, "failed to validate challenge response"), http.StatusBadRequest)
+		api.WriteError(w, errors.AddContext(err, "failed to parse request body"), http.StatusBadRequest)
 		return
 	}
-	// Validate the email address.
-	email := req.PostFormValue("email")
-	e, err := mail.ParseAddress(email)
-	if err != nil {
+	parsed, err := mail.ParseAddress(payload.Email)
+	if err != nil || payload.Email != parsed.Address {
```

Comment:
This is quite odd? I don't think I follow. So you parse the address from the `payload.Email` field and then later on you error out when `payload.Email` is not equal to `parsed.Address`. Doesn't it make sense to alter `ParseAddress` and return an `error` in that case?

Reply:
That was my initial approach but it comes with its own problems. This approach guarantees that unless the email is valid and simple (e.g. "joe@example.com" and not "Joe Smith <joe@example.org>") the process will stop with an error.

Reply:
Well it guarantees this as long as you remember to consider `payload.Email != parsed.Address` as an error. I'll approve but I don't understand why this condition can't be inside of `ParseAddress`.

Reply:
ParseAddress is a stdlib function. I'll add this to the F/U list.